### PR TITLE
Fix the JPAKE driver interface for user+peer

### PR DESCRIPTION
**Pull request by mprse**
## Description

Support any j-pake user/peer values in core.
Driver/built-in implementation still has limitation to accept only "client" or "server" values for j-pake user/peer.
Adapted and extended tests.

Resolves https://github.com/Mbed-TLS/mbedtls/issues/7289

## Gatekeeper checklist

- [x] **changelog** provided
- [ ] **backport** done, or not required
- [x] **tests** provided

## Notes for the submitter

Please refer to the [contributing guidelines](https://github.com/Mbed-TLS/mbedtls/blob/development/CONTRIBUTING.md), especially the
checklist for PR contributors.

### CODE REVIEW

File: library/psa_crypto_pake.c [hunk 1]
```diff
@@ -541,7 +594,7 @@ psa_status_t mbedtls_psa_pake_abort(mbedtls_psa_pake_operation_t *operation)
 
 #if defined(MBEDTLS_PSA_BUILTIN_ALG_JPAKE)
     if (operation->alg == PSA_ALG_JPAKE) {
-        operation->role = PSA_PAKE_ROLE_NONE;
+        operation->role = 0;
```
Comment (AndrzejKurek) [inline library/psa_crypto_pake.c:597]: This will now clash with `MBEDTLS_ECJPAKE_CLIENT`

File: library/psa_crypto_pake.c [hunk 2]
```diff
@@ -216,7 +223,20 @@ psa_status_t mbedtls_psa_pake_setup(mbedtls_psa_pake_operation_t *operation,
 
     operation->password = mbedtls_calloc(1, password_len);
     if (operation->password == NULL) {
-        return PSA_ERROR_INSUFFICIENT_MEMORY;
+        status = PSA_ERROR_INSUFFICIENT_MEMORY;
+        goto error;
```
Comment (AndrzejKurek) [inline library/psa_crypto_pake.c:227]: I believe that a `goto` is not needed at this point, but it should be harmless.

File: library/psa_crypto_pake.c [hunk 3]
```diff
@@ -238,7 +270,27 @@ psa_status_t mbedtls_psa_pake_setup(mbedtls_psa_pake_operation_t *operation,
             goto error;
         }
 
-        operation->role = role;
+        const size_t user_peer_len = sizeof(jpake_client_id); // client and server have the same length
```
Comment (AndrzejKurek) [inline library/psa_crypto_pake.c:273]: That's such a small optimization that I don't know if I wouldn't suggest just having two lengths for the sake of being more readable. If the ID's change then it's a hidden optimization that's not obvious.

File: library/psa_crypto_pake.c [hunk 4]
```diff
@@ -238,7 +270,27 @@ psa_status_t mbedtls_psa_pake_setup(mbedtls_psa_pake_operation_t *operation,
...
+        if (memcmp(user, jpake_server_id, actual_user_len) == 0 &&
+            memcmp(peer, jpake_client_id, actual_peer_len) == 0) {
+            operation->role = MBEDTLS_ECJPAKE_SERVER;
+        } else {
+            status = PSA_ERROR_NOT_SUPPORTED;
+            goto error;
+        }
+
+        /* Role has been set, release user/peer buffers. */
+        mbedtls_free(user); mbedtls_free(peer);
```
Comment (AndrzejKurek) [inline library/psa_crypto_pake.c:293]: Potential double free if `psa_pake_ecjpake_setup` fails. Please move it to before `return PSA_SUCCESS;`.

Comment (AndrzejKurek) [reply library/psa_crypto_pake.c:293]: It would also be good to have such coverage if it's easy to achieve.

**Review by AndrzejKurek — CHANGES_REQUESTED** [review]
Some changes requested.

File: library/psa_crypto_pake.c (hunk 2 above)
Comment (mprse) [reply library/psa_crypto_pake.c:227]: We could have returned here, but I think that both versions are acceptable.

File: library/psa_crypto_pake.c (hunk 3 above)
Comment (mprse) [reply library/psa_crypto_pake.c:273]: With two lengths I would have to add two additional conditions to the if statement.

File: tests/suites/test_suite_psa_crypto_pake.function [hunk 5]
```diff
@@ -1139,37 +1108,43 @@ void pake_input_getters_peer()
...
+        PSA_ASSERT(psa_pake_abort(&operation));
 
-    TEST_EQUAL(psa_crypto_driver_pake_get_peer_len(&operation.data.inputs, &peer_len_ret),
-               PSA_ERROR_BAD_STATE);
+        PSA_ASSERT(psa_pake_setup(&operation, &cipher_suite));
 
-    PSA_ASSERT(psa_pake_set_peer(&operation, peer, peer_len));
+        TEST_EQUAL(psa_crypto_driver_pake_get_peer(&operation.data.inputs,
+                                                   (uint8_t *) &peer_ret,
+                                                   10, &buffer_len_ret),
```
Comment (AndrzejKurek) [inline tests/suites/test_suite_psa_crypto_pake.function:1121]: Pre-existing: hardcoded 10

**Review by AndrzejKurek — APPROVED** [review]
Looks good.

**Comment by mpg** [comment]
@AndrzejKurek @davidhorstmann-arm @gilles-peskine-arm @mprse The checkbox for backport is not checked and no reason is given. I'm fixing this but letting you all know (author, both reviewers, gatekeeper) as everyone should have checked this.

**Comment by gilles-peskine-arm** [comment]
Sorry, what's the problem here? This is about a feature that doesn't exist in 2.28, so the reason for not backporting is obvious. The only thing that's missing is an explicit statement that there is no incidental bug fix alongside the main purpose of the PR, but if there is one, the author and the reviewers also missed it.

**Comment by AndrzejKurek** [comment]
I think that the problem exists in the context of release preparation, as scripts / automatic tools will pick it up and show as a problem, whereas checking the box and giving an explanation will make it so that the person preparing the release won't have to investigate further and ponder.

**Comment by mpg** [comment]
The problem is the box was not ticked and the default text was still present, it was looking like this before my edit:

- [ ] **backport** done, or not required

**Pull request merged**